File: Platform/Database/Allors.Database.Protocol/Decompressor.cs
// <copyright file="Compressor.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Protocol
{
    using System.Collections.Generic;

    public class Decompressor
    {
        private readonly Dictionary<string, string> valueByKey;

        public Decompressor() => this.valueByKey = new Dictionary<string, string>();

        public string Read(string compressed, out bool first)
        {
            first = false;

            if (!string.IsNullOrEmpty(compressed))
            {
                if (compressed[0] == Compressor.IndexMarker)
                {
                    first = true;
                    var secondIndexMarkerIndex = compressed.IndexOf(Compressor.IndexMarker, 1);
                    var key = compressed.Substring(1, secondIndexMarkerIndex - 1);
                    var value = compressed.Substring(secondIndexMarkerIndex + 1);
                    this.valueByKey.Add(key, value);
                    return value;
                }

                return this.valueByKey[compressed];
            }

            return null;
        }
    }
}

File: Core/Workspace/Typescript.legacy/Site/wwwroot/Core/Workspace/Protocol/Decompressor.ts
namespace Allors.Protocol {
  import MetaPopulation = Meta.MetaPopulation;

  export const createMetaDecompressor = (decompressor: Decompressor, metaPopulation: MetaPopulation) => (compressed: string) => {
        return metaPopulation.metaObjectById.get(decompressor.read(compressed, v => { }));
    };

    export class Decompressor {
        private valueByKey: Map<string, string>;

        constructor() {
            this.valueByKey = new Map();
        }

        public read(compressed: string, first: (value: string) => void): string {

            if (compressed !== undefined && compressed !== null) {
                if (compressed[0] === Compressor.indexMarker) {
                    const secondIndexMarkerIndex = compressed.indexOf(Compressor.indexMarker, 1);
                    const key = compressed.slice(1, secondIndexMarkerIndex);
                    const value = compressed.slice(secondIndexMarkerIndex + 1);
                    this.valueByKey.set(key, value);
                    first(value);
                    return value;
                }

                return this.valueByKey.get(compressed);
            }

            return null;
        }
    }
}
